### PR TITLE
Read normalized cache

### LINKED ISSUE #266: Add read support for the normalized cache
This should be able to closely follow the pattern established by the HttpCache, but we won't be able to use OkHttp interceptors because we need to convert from request -> data. Not request -> responseBody.

In terms of the actual reading we can follow the [iOS approach](https://github.com/apollographql/apollo-ios/blob/master/Sources/GraphQLResultReader.swift#L1) and use `GraphQLResolver`, which is essentially a function field -> Object. For network this function would just access a buffered json map with the field name. For normalized cache we would look up the field and follow references.

Another option, would be to have a `CacheResponseReader` and leave `BufferedResponseReader` as is.

### PR DESCRIPTION
**Pull request by sav007**
Introduces reading from cache.

Closes #266

### CODE REVIEW

File: apollo-integration/src/test/java/com/apollographql/android/impl/NormalizedCacheTestCase.java [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    server = new MockWebServer();
+
+    OkHttpClient okHttpClient = new OkHttpClient.Builder().build();
+    cacheStore = new InMemoryCacheStore();
+
+    apolloClient = ApolloClient.builder()
+        .serverUrl(server.url("/"))
+        .okHttpClient(okHttpClient)
+        .normalizedCache(cacheStore, new CacheKeyResolver() {
+          @Nullable @Override public String resolve(Map<String, Object> jsonObject) {
```
Comment (digitalbuddha) [inline apollo-integration/src/test/java/com/apollographql/android/impl/NormalizedCacheTestCase.java]: this should not be nullable

File: apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloCall.java [hunk 2]
```diff
@@ -24,32 +29,42 @@
...
   }
 
   @Nonnull @Override public Response<T> execute() throws IOException {
     synchronized (this) {
       if (executed) throw new IllegalStateException("Already Executed");
       executed = true;
     }
+
+    Response<T> cachedResponse = cachedResponse();
+    if (cachedResponse != null) {
```
Comment (digitalbuddha) [inline apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloCall.java]: so not part of this pr but it kinda makes me sad that we have null everywhere in api. It would be nice to do something like `Optional`, I wouldn't mind doing the refactor myself if no one has issues. Also open to other suggestions @BenSchwab @sav007 @brianPlummer

File: apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloCall.java [hunk 3]
```diff
@@ -132,17 +157,53 @@ private RealApolloCall(Operation operation, HttpUrl serverUrl, Call.Factory http
...
+      return null;
+    }
+
+    T cachedData = cachedData();
+    if (cachedData != null) {
+      return new Response<>(operation, cachedData, null);
+    }
+
+    if (cacheControl == CacheControl.CACHE_ONLY) {
+      return new Response<>(operation, null, null);
```
Comment (digitalbuddha) [inline apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloCall.java]: pls use constants that represent what null means, its difficult to see if this is correct or not when there are 2 unnamed params of null null.  Also why did ew not write the whole thing in kotlin? :-) too late now I guess

File: apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloCall.java [hunk 4]
```diff
@@ -132,17 +157,53 @@ private RealApolloCall(Operation operation, HttpUrl serverUrl, Call.Factory http
...
+    }
+
+    return null;
+  }
+
+  @SuppressWarnings("unchecked") private T cachedData() {
```
Comment (digitalbuddha) [inline apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloCall.java]: @Nullable pls

File: apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloCall.java (hunk 3 above)
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloCall.java]: Kotlin I guess because it will bring runtime with it, here we use pure java. But I know this is would be nicer to have on Kotlin. Maybe we need to consider some `Optional` approach instead of null

File: apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloCall.java (hunk 2 above)
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloCall.java]: yeah, agree we need to consider using `Optional`

File: apollo-integration/src/test/java/com/apollographql/android/impl/NormalizedCacheTestCase.java (hunk 1 above)
Comment (sav007) [inline apollo-integration/src/test/java/com/apollographql/android/impl/NormalizedCacheTestCase.java]: that how the we ported from iOS I guess. if null that means we are going to use path as a key.
We can change `CacheKeyResolver` to return not String but some class like `CacheKey`. And null replace with `CacheKey.EMPTY` smth like this

File: apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloCall.java (hunk 2 above)
Comment (BenSchwab) [inline apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloCall.java]: +1 at the very least we can use the @Nullable so we can get lint warnings

File: apollo-integration/src/test/java/com/apollographql/android/impl/NormalizedCacheTestCase.java (hunk 1 above)
Comment (sav007) [inline apollo-integration/src/test/java/com/apollographql/android/impl/NormalizedCacheTestCase.java]: will create issue for this

File: apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloCall.java (hunk 2 above)
Comment (digitalbuddha) [inline apollo-runtime/src/main/java/com/apollographql/android/impl/RealApolloCall.java]: https://github.com/apollographql/apollo-android/issues/290 Feel free to do it otherwise I'll do it after the sql stuff

File: apollo-runtime/src/main/java/com/apollographql/android/cache/normalized/Cache.java [hunk 5]
```diff
@@ -49,19 +59,15 @@ public ResponseNormalizer responseNormalizer() {
...
-
     }
-
   }, new CacheKeyResolver() {
     @Nullable @Override public String resolve(Map<String, Object> jsonObject) {
       return null;
     }
   });
 
   //Todo: add interceptor for reading (https://github.com/apollographql/apollo-android/issues/266)
```
Comment (BenSchwab) [inline apollo-runtime/src/main/java/com/apollographql/android/cache/normalized/Cache.java]: Remove this Todo?

File: apollo-integration/src/test/java/com/apollographql/android/impl/NormalizedCacheTestCase.java (hunk 1 above)
Comment (sav007) [inline apollo-integration/src/test/java/com/apollographql/android/impl/NormalizedCacheTestCase.java]: https://github.com/apollographql/apollo-android/issues/291
